### PR TITLE
DrivAerML: 10-epoch linear warmup + cosine T_max=30

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

10-epoch warmup provides a more gradual ramp than 5-epoch. AB-UPT uses Lion with a warmup+cosine schedule for DrivAerML.

## Instructions

Add a 10-epoch linear warmup before cosine annealing:
- Implement `SequentialLR([LinearLR(optimizer, start_factor=0.1, end_factor=1.0, total_iters=10), CosineAnnealingLR(optimizer, T_max=30)], milestones=[10])`
- Add `--warmup-epochs 10` flag

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --warmup-epochs 10 \
  --wandb_group dm-warmup-sweep
```

## Reporting Requirements
- Best val + test `surface_rel_l2_pct`
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chrome:

## Results

### 10-epoch linear warmup + cosine T_max=30 (DrivAerML)

| Metric | Baseline | This Experiment |
|--------|----------|----------------|
| val best `surface_rel_l2_pct` | **3.997%** @ ep467 | **6.225%** @ ep105 |
| test `surface_rel_l2_pct` | 6.244% (older config) | N/A (diverged before test eval) |
| External targets | AB-UPT 3.82% \| Transolver-3 3.71% | — |

**W&B run ID:** `w1c7l3jo`
**Peak memory:** ~6.0 GB (GPU 0)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=999 SENPAI_TIMEOUT_MINUTES=360 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --warmup-epochs 10 \
  --wandb-group dm-warmup-sweep \
  --agent chrome --wandb-name "chrome/dm-warmup-10ep-cosine30-v2"
```

### What happened

The 10-epoch linear warmup (start_factor=0.1 → 1.0) was implemented using `SequentialLR([LinearLR, CosineAnnealingLR])`. The scheduler steps per batch (not per epoch), so warmup iters were scaled to `warmup_epochs * steps_per_epoch = 10 * 394 = 3940` steps.

**Initial run (v1)** had a bug: warmup `total_iters` and milestone were set to epoch counts (10) instead of batch-step counts. Since the scheduler steps per batch, the warmup was consumed in the first 10 batches of epoch 1, and the cosine oscillated wildly — training diverged after epoch ~130. V1 best val: 15.02%.

**Fixed run (v2)** correctly scaled warmup to batch steps. Training converged well:
- Warmup epochs 1–10: smooth LR ramp from 5e-5 → 5e-4, val improved from 63% → 24%
- Cosine phase epochs 11–105: steady improvement to 6.225% (new bests at every ~10 epochs)
- **Catastrophic divergence at epoch 114:** val spiked from 10.8% → 81.9%, grad norms from 0.4 → 8.8. Training never recovered — oscillated chaotically between 30–52% from ep115–124 before I killed it.

The divergence is likely caused by the combination of: (1) batch_size=1 with no gradient clipping (`grad_clip=0.0`), and (2) the rapid cosine oscillation (T_max=30 in batch-steps = ~0.076 epochs per half-cycle). A single bad gradient update at a LR peak can push the model into an unstable regime.

**Verdict: The warmup itself worked correctly and showed good early convergence, but the result (6.225%) does not beat the baseline (3.997%).** The run was cut short by divergence at epoch 114 — the baseline reached 3.997% at epoch 467, so this run didn't have enough stable training time. The warmup may still be beneficial if combined with gradient clipping to prevent the catastrophic divergence.

### Suggested follow-ups
- **Add gradient clipping** (e.g. `--grad-clip 1.0`) to prevent catastrophic divergence events — this would let warmup runs train stably to the 400+ epochs needed to reach baseline territory
- **Increase cosine T_max** — T_max=30 in batch-steps gives very rapid cycling (~13 full cosine periods per epoch). A larger T_max might give smoother training dynamics
- Try **CosineAnnealingWarmRestarts** with proper per-epoch stepping for more controlled LR decay

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3070 because ADVISOR: Closing — 10-epoch warmup without gc diverged at ep114. Best val 6.225% at ep105 (+56% above 3.997% baseline). The warmup itself worked correctly (63%→24% in epochs 1-10), but the run still hit catastrophic gradient explosion at a cosine LR peak. Without gc, the warmup contribution is overwhelmed by the same instability that kills all DM non-champion configs. The companion #3069 (5-epoch warmup) likely shows the same pattern. The real test is warmup + gc combined (historia #3119 tests higher LR + 20-epoch warmup). Good diagnosis, chrome..
